### DISCUSSION TITLE
http2 h2/c Support

### DISCUSSION THREAD

**Discussion by mario4rma** [directus/directus#20022] [Feature Requests]
### Summary

The native support of h2 and or h2/c would be a great benefit. For performance and the choice of infrastructure.

### Motivation

There are some limitations in certain infrastructures. The best example is, when you are using CloudRun inside a GCP.
You have a hard request limit defined by google CloudRun [https://cloud.google.com/run/quotas](https://cloud.google.com/run/quotas). You are limited to 32MB per Request unless you use http2.

The problem is, that as soon as you activate the http2 feature for your CloudRun container, the communcation breaks between the container and the directus instance. This is because CloudRun does not migrate h2 requests down to h1.1 requests anymore. 

In fact you are not able to send requests (upload files) which are larger than 32MB. 

### Detailed Design

We should be using [https://nodejs.org/api/http2.html](https://nodejs.org/api/http2.html) Furthermore the server needs to support h2c [https://http2.github.io/faq/#can-i-implement-http2-without-implementing-http11](https://http2.github.io/faq/#can-i-implement-http2-without-implementing-http11) as the CloudRun instance will terminate TLS.

### Must Have:
- Support for http2 Requests
- Support for h2c Requests

### Should Have:
- A parameter in the config to load TLS key Files

### Could Have:
- A flag in the config to choose between http1.1 and http2 server

### Won't Have:
- 

### Drawbacks

If there is a flag to choose between http1.1 and h2 server there should be no breaking changes for the users.
Although the switch to http2 will need some configs in their infrastructure

### Alternatives

An alternative would be that the file upload endpoint can handle uploads in chunks. Especially when using the directus SDK

### Adoption Strategy

If it is handled by the config (choose http2 over http1.1), this will be no breaking change and there is no migration needed.
It is the decision of the developers if they want to switch to h2.

**Comment by keesvanbemmel** [comment]
This would be a great addition for us as we exclusively run Directus on GCR.

File uploads has been fixed with the newly released TUS support, but for larger downloads we would still neet http2 support.

**Comment by sufyman** [comment]
would be great!

**Comment by amenabe22** [comment]
Any progress on this?
